File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarFirmwareUpdateApi.kt
// Copyright © 2024 Polar Electro Oy. All rights reserved.
package com.polar.sdk.api

import com.polar.sdk.api.model.FirmwareUpdateStatus
import io.reactivex.rxjava3.core.Flowable
import com.polar.sdk.api.model.PolarFirmwareVersionInfo

/**
 * Polar firmware update API.
 *
 * Requires feature [FEATURE_POLAR_FIRMWARE_UPDATE]
 *
 */
interface PolarFirmwareUpdateApi {
    /**
     * Updates firmware to given device from local file.
     *
     * @param identifier Polar device ID or BT address
     * @param filePath path to the file
     * @return [Flowable] emitting status of firmware update
     */
    fun updateFirmwareLocal(identifier: String, filePath: String): Flowable<FirmwareUpdateStatus>

    /**
     * Updates firmware to given device.
     *
     * @param identifier Polar device ID or BT address
     * @return [Flowable] emitting status of firmware update
     */
    fun updateFirmware(identifier: String): Flowable<FirmwareUpdateStatus>

    fun getFirmwareInfo(identifier: String): PolarFirmwareVersionInfo?
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/PolarFirmwareUpdateApi.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation
import RxSwift

public protocol PolarFirmwareUpdateApi {
    /**
     * Checks firmware update to given device.
     *
     * - Parameter identifier: Polar device ID or BT address
     * - Returns: Observable emitting status of firmware update check
     */
    func checkFirmwareUpdate(_ identifier: String) -> Observable<CheckFirmwareUpdateStatus>
  
    /**
      * Updates firmware to given device.
      *
      * - Parameter identifier: Polar device ID or BT address
      * - Parameter filePath: path to firmware file
      * - Parameter version: firmware version
      * - Returns: Observable emitting status of firmware update
      */
    func updateFirmwareLocal(_ identifier: String, _ filePath: String, _ version: String) -> Observable<FirmwareUpdateStatus>
    /**
     * Updates firmware to given device.
     *
     * - Parameter identifier: Polar device ID or BT address
     * - Returns: Observable emitting status of firmware update
     */
    func updateFirmware(_ identifier: String) -> Observable<FirmwareUpdateStatus>
    
    /**
     * Updates firmware from specific URL
     *  - Parameter identifier: Polar device ID or BT address
     *  - Parameter fromFirmwareURL: URL to firmware file. Firmware file must be compatible with target device.
     *  - Returns: Observable emitting status of firmware update
     */
    func updateFirmware(_ identifier: String, fromFirmwareURL: URL) -> Observable<FirmwareUpdateStatus>
  
    func getFirmwareInfo(_ identifier: String) -> PolarFirmwareVersionInfo?
}
